app/models/source.rb

```diff
@@ -47,4 +49,8 @@ def remove_availability_status!(source = nil)
     remove_availability_status(source)
     save!
   end
+
+  def validate_no_applications
+    raise(ActiveRecord::RecordNotDestroyed, "Applications must be removed before destroying parent source") if applications.any?
```

Comment:
isn't it redundant? Postgres includes Foreign key on applications:sources_id so it's deleted by database

Reply:
It is automatically deleted by cascade yes - this is just to prevent a potential race condition where a Superkey Source is deleted and if there are any application resources there could be a case where the cascade delete happens too fast and the worker can't pick up the credentials in order to clean up the resources in AWS. 

This hasn't bit us yet, but that's only because the worker basically picks up the message and gets the credential right away. If there is kafka lag or something else that makes it take longer than a split second it would fail to delete the resources.

Reply:
Ah, get it